blog/content/post/2019-07-24-esp8266-quick-guide-for-embedded-c-developers.md

```diff
@@ -0,0 +1,225 @@
+we'll based on **NodeMCU** development board with **ESP-12** module. The board
+has a USB-UART converter, voltage regulator, buttons and gold pin headers. There
+is no need for any soldering. It's a really convenient solution.
+
+![nodemcu](/img/nodemcu.png)
+
+## Get started
+
+Espressif provides two types of SDK:
+
+* **RTOS SDK** - based on FreeRTOS,
+
+* **Non-OS SDK** - based on timers and callbacks.
+
+More details about those SDKs you can find in Chapter **1.3. ESP8266 SDK** of
+[**ESP8266 SDK - Getting Started Guide**](https://www.espressif.com/sites/default/files/documentation/2a-esp8266-sdk_getting_started_guide_en.pdf).
+Good alternative for **Espressif RTOS SDK** is
+[**ESP-Open-RTOS**](https://github.com/SuperHouse/esp-open-rtos). In this post,
+however, we will focus on the Non-OS SDK. If you are interested in
+the RTOS SDK please let us know. We will try to prepare a post on this topic as
+well.
+
+### Preparing the environment
+
+Before we go to the code example we have to prepare environment for work.
+Things we need are:
+
+* Xtensa lx106 architecture toolchain,
+* ESP8266 SDK published by vendor,
+* tool for flashing the on-chip memories.
+
+Of course, we can assemble all the tools manually, but
+**it's not the optimal solution**. Instead, we can use
+[**ESP-Open-SDK**](https://github.com/pfalcon/esp-open-sdk), which contains
+**everything we need in one place**. If you also prefer this solution, the only
+thing you have to do is to execute those simple steps:
+
+* clone repository:
+    
+    ```
+    git clone --recursive https://github.com/pfalcon/esp-open-sdk.git
+    ```
+
+* install dependencies:
+
+    ```
+    sudo apt-get install make unrar-free autoconf automake libtool gcc g++ gperf \
+    flex bison texinfo gawk ncurses-dev libexpat-dev python-dev python python-serial \
+    sed git unzip bash help2man wget bzip2 libtool-bin
+    ```
+
+* build the toolchain:
+
+    We can build the toolchain in two ways:
+    - with a completely **standalone** ESP8266 SDK containing the vendor SDK
+files merged into the toolchain
+    - with a **separate** toolchain and vendor SDK files
+
+    We definitely **prefer the second option**. It will be needed to add
+**extra -I and -L flags** during compilation, but this approach gives us some
+benefits. We'll have a **better control** in what we do and we'll can easily use
+different version of sdk when needed. To build the separate version all you need
+is to use make with `STANDALONE=n`.
+
+    ```
+    make STANDALONE=n
+    ```
+
+    It'll take few minutes, so be patient. When the compilation process is
+completed, the toolchain will be available in the `xtensa-lx106-elf/`
+subdirectory. At the end of build process, command for adding the
+`xtensa-lx106-elf/bin/` subdirectory to your **PATH environment variable** will
+be shown. You can **save it**, because it will be needed execute it each time
+you want to use the **xtensa-lx106-elf-gcc** and other tools. In my case,
```

Comment:
each time means one time per programming session?

Reply:
Yes, I'll precise it.